Userland/Libraries/LibMarkdown/Text.h

```diff
@@ -1,54 +1,129 @@
  */
 
 #pragma once
 
 #include <AK/Noncopyable.h>
+#include <AK/NonnullOwnPtrVector.h>
+#include <AK/OwnPtr.h>
 #include <AK/String.h>
-#include <AK/Vector.h>
 
 namespace Markdown {
 
 class Text final {
-    AK_MAKE_NONCOPYABLE(Text);
-
 public:
-    struct Style {
-        bool emph { false };
-        bool strong { false };
-        bool code { false };
-        String href;
-        String img;
+    struct Node {
+        virtual void render_to_html(StringBuilder& builder) const = 0;
+        virtual void render_for_terminal(StringBuilder& builder) const = 0;
+        virtual size_t length() const = 0;
+
+        virtual ~Node() { }
+    };
+
+    struct Emph : Node {
+        bool strong;
+        NonnullOwnPtr<Node> child;
+        Emph(bool strong, NonnullOwnPtr<Node> child)
+            : strong(strong)
+            , child(move(child))
+        {
+        }
+
+        virtual void render_to_html(StringBuilder& builder) const override;
+        virtual void render_for_terminal(StringBuilder& builder) const override;
+        virtual size_t length() const override;
     };
 
-    struct Span {
+    struct Code : Node {
+        NonnullOwnPtr<Node> code;
+
+        Code(NonnullOwnPtr<Node> code)
+            : code(move(code))
+        {
+        }
+
+        virtual void render_to_html(StringBuilder& builder) const override;
+        virtual void render_for_terminal(StringBuilder& builder) const override;
+        virtual size_t length() const override;
+    };
+
+    struct TextNode : Node {
         String text;
-        Style style;
+
+        TextNode(StringView const& text)
+            : text(text)
+        {
+        }
+
+        virtual void render_to_html(StringBuilder& builder) const override;
+        virtual void render_for_terminal(StringBuilder& builder) const override;
+        virtual size_t length() const override;
+    };
+
+    struct LinkNode : Node {
+        bool is_image;
+        NonnullOwnPtr<Node> text;
+        NonnullOwnPtr<Node> href;
+
+        LinkNode(bool is_image, NonnullOwnPtr<Node> text, NonnullOwnPtr<Node> href)
+            : is_image(is_image)
+            , text(move(text))
+            , href(move(href))
+        {
+        }
+
+        virtual void render_to_html(StringBuilder& builder) const override;
+        virtual void render_for_terminal(StringBuilder& builder) const override;
+        virtual size_t length() const override;
     };
 
-    explicit Text(String&& text);
-    Text(Text&& text) = default;
-    Text() = default;
+    struct MultiNode : Node {
+        NonnullOwnPtrVector<Node> children;
 
-    Text& operator=(Text&&) = default;
+        virtual void render_to_html(StringBuilder& builder) const override;
+        virtual void render_for_terminal(StringBuilder& builder) const override;
+        virtual size_t length() const override;
+    };
+
+    struct Token {
+        String data;
+        bool left_flanking;
+        bool right_flanking;
+        bool is_run;
```

Comment:
`Token`s are only used internally by `Text` during the parsing process. I suggest moving the `struct` definition at least to the private section.

Or even better, if you wanna invest the effort: In the private section, do only a struct declaration (`struct Token;`), and define (`struct Text::Token { … }`) in the file `Text.cpp`. This way, the internals are completely hidden from any users of `Markdown::Text`.

Reply:
Tokens are now private.